protocol/messenger_status_updates.go

```diff
@@ -0,0 +1,247 @@
+		m.logger.Debug("user status should not be broadcasted")
+		return
+	}
+
+	currStatus, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return
+	}
+
+	if err := m.sendUserStatus(*currStatus); err != nil {
+		m.logger.Debug("Error when sending the latest user status", zap.Error(err))
+	}
+}
+
+func (m *Messenger) sendCurrentUserStatusToCommunity(community *communities.Community) error {
+	shouldBroadcastUserStatus, err := m.settings.ShouldBroadcastUserStatus()
+	if err != nil {
+		return err
+	}
+
+	if !shouldBroadcastUserStatus {
+		m.logger.Debug("user status should not be broadcasted")
+		return nil
+	}
+
+	status, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return err
+	}
+
+	status.Clock = uint64(time.Now().Unix())
+
+	err = m.settings.SaveSetting("current-user-status", status)
+	if err != nil {
+		return err
+	}
+
+	statusUpdate := &protobuf.StatusUpdate{
+		Clock:      status.Clock,
+		StatusType: protobuf.StatusUpdate_StatusType(status.StatusType),
+		CustomText: status.CustomText,
+	}
+
+	encodedMessage, err := proto.Marshal(statusUpdate)
+	if err != nil {
+		return err
+	}
+
+	rawMessage := common.RawMessage{
+		LocalChatID:         community.StatusUpdatesChannelID(),
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_STATUS_UPDATE,
+		ResendAutomatically: true,
+	}
+
+	_, err = m.sender.SendPublic(context.Background(), rawMessage.LocalChatID, rawMessage)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Messenger) broadcastLatestUserStatus() {
+	m.logger.Debug("broadcasting user status")
+	m.sendCurrentUserStatus()
+	go func() {
+		for {
+			select {
+			case <-time.After(5 * time.Minute):
+				m.sendCurrentUserStatus()
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+}
+
+func (m *Messenger) SetUserStatus(newStatus int, newCustomText string) error {
+	if len([]rune(newCustomText)) > maxStatusMessageText {
+		return fmt.Errorf("custom text shouldn't be longer than %d", maxStatusMessageText)
+	}
+
+	if newStatus != int(protobuf.StatusUpdate_ONLINE) && newStatus != int(protobuf.StatusUpdate_DO_NOT_DISTURB) {
+		return fmt.Errorf("unknown status type")
+	}
+
+	currStatus, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return err
+	}
+
+	if newStatus == currStatus.StatusType && newCustomText == currStatus.CustomText {
+		m.logger.Debug("Status type did not change")
+		return nil
+	}
+
+	currStatus.StatusType = newStatus
+	currStatus.CustomText = newCustomText
+
+	return m.sendUserStatus(*currStatus)
+}
+
+func (m *Messenger) HandleStatusUpdate(state *ReceivedMessageState, statusMessage protobuf.StatusUpdate) error {
+	if err := ValidateStatusUpdate(statusMessage); err != nil {
+		return err
+	}
+
+	currentStatus, err := m.GetCurrentUserStatus()
```

Comment:
This looks like can go inside the next `if` block, so we save a database query in the most likely scenario (i.e the status is not coming from us)